### PR TITLE
core.bitops: Add bitwise rotate functions

### PR DESCRIPTION
**Pull request by CounterPillow**
Adds functions for a bitwise rotate left and rotate right operation. This operation is common enough that some architectures such as x86 have it as its own instruction; while many compilers will detect the used pattern and optimise it down to one instruction, some compilers (e.g. DMD) won't do so, in which case having this as a part of druntime allows for one hand-written assembly implementation later on.

The choice of `uint` as the type for the number of shifts is a bit arbitrary; any unsigned integer type would do, as the number of places to shift is usually fairly small (and wraps at `T.sizeof` anyway).

### CODE REVIEW

**Comment by ibuclaw** [comment]
Yep, I can train gdc to recognise and rewrite this as an lrotate or rrotate expression. :+1:

**Comment by CounterPillow** [comment]
Any opinions on `isArithmetic` being used instead of `isIntegral`? I don't know whether there are valid use cases for bit rotating a floating point value (I'm not too familiar with bit twiddling hacks). Is there any downside to allowing floating point values to be bit rotated?

Also, can anybody think of more edge cases to add to the unit tests? Especially the `cast(T)` is making me nervous (and it was required to get types shorter than `int` working).

File: src/core/bitop.d [hunk 1]
```diff
@@ -706,3 +706,38 @@ version (D_InlineAsm_X86_64)
...
+/**
+ *  Bitwise rotates unsigned arithmetic types
+ */
+pure T rotl(T)(in T u, in uint s)
+    if (__traits(isArithmetic, T) && __traits(isUnsigned, T))
+{
+    return cast(T) ((u << s) | (u >> (-s & (T.sizeof * 8 - 1))));
+}
+
+pure T rotr(T)(in T u, in uint s)
```
Comment (MartinNowak) [inline src/core/bitop.d]: Use `/// ditto` if both functions should share the same documentation.

File: src/core/bitop.d [hunk 2]
```diff
@@ -706,3 +706,38 @@ version (D_InlineAsm_X86_64)
...
+    if (__traits(isArithmetic, T) && __traits(isUnsigned, T))
+{
+    return cast(T) ((u >> s) | (u << (-s & (T.sizeof * 8 - 1))));
+}
+
+unittest
```
Comment (MartinNowak) [inline src/core/bitop.d]: Make this a [documented unittest](http://dlang.org/spec/unittest.html#documented-unittests).

**Comment by MartinNowak** [comment]
The instructions are called `ror` and `rol` by Intel, and `ror` by ARM, let's stick with that.

File: src/core/bitop.d [hunk 3]
```diff
@@ -706,3 +706,38 @@ version (D_InlineAsm_X86_64)
...
     }
 }
+
+/**
+ *  Bitwise rotates unsigned arithmetic types
+ */
+pure T rotl(T)(in T u, in uint s)
+    if (__traits(isArithmetic, T) && __traits(isUnsigned, T))
+{
+    return cast(T) ((u << s) | (u >> (-s & (T.sizeof * 8 - 1))));
```
Comment (MartinNowak) [inline src/core/bitop.d]: What happens when `s` is bigger than `8 * T.sizeof`? If that behavior is inconsistent across architectures, an `assert(s < 8 * T.sizeof)` would be adequate.

File: src/core/bitop.d [hunk 4]
```diff
@@ -706,3 +706,38 @@ version (D_InlineAsm_X86_64)
         }
     }
 }
+
+/**
+ *  Bitwise rotates unsigned arithmetic types
+ */
+pure T rotl(T)(in T u, in uint s)
```
Comment (MartinNowak) [inline src/core/bitop.d]: I'd prefer `n` or `count` instead of `s`, and `val` or `value` instead of `u`.
Any particular reason for naming them `u` and `s`?

**Comment by MartinNowak** [comment]
There should also be an overload that takes the shift number as template parameter, so more efficient instructions w/ immediate argument can be used.

File: src/core/bitop.d [hunk 5]
```diff
@@ -706,3 +706,38 @@ version (D_InlineAsm_X86_64)
...
+    if (__traits(isArithmetic, T) && __traits(isUnsigned, T))
```
Comment (MartinNowak) [inline src/core/bitop.d]: Bit operations on floats are only possible after reinterpreting the representation to an integer, so please `isIntegral` here.

Comment (MartinNowak) [inline src/core/bitop.d]: There are no unsigned floats btw.

Comment (ibuclaw) [inline src/core/bitop.d]: Wouldn't rotate only work for integers?

**Comment by CounterPillow** [comment]
Pushed a new commit to address the raised concerns. I'll squash them when it's ready to merge. Time for round 2 of suggestions.

Does anybody happen to know whether it's allowed to use std.exception in druntime unittests to get `assertThrown`? That way I could include a couple more unittests.

**Comment by MartinNowak** [comment]
> Does anybody happen to know whether it's allowed to use std.exception in druntime unittests to get assertThrown? That way I could include a couple more unittests.

No, it's not possible to use phobos in druntime.

**Comment by MartinNowak** [comment]
I've slightly tweaked the documentation, see af2fa26b1434ae82c617ea554bcc7bb5c49a32bb.

File: src/core/bitop.d [hunk 6]
```diff
@@ -706,3 +706,55 @@ version (D_InlineAsm_X86_64)
         }
     }
 }
+
+/**
+ *  Bitwise rotates unsigned integer types
```
Comment (MartinNowak) [inline src/core/bitop.d]: It's better to use `rotate` here, b/c the function doesn't do anything on its own account.

**Comment by CounterPillow** [comment]
I see, you merged it right away without squashing the commits, though I guess the commit history isn't too bad with just two commits.

Also, I forgot to add my name to the authors list on the top, though I'm not sure if there needs to be some minimum required amount of contribution to the module before that's put there.
